src/plugins/hiddenColumns/hiddenColumns.js

```diff
@@ -436,105 +391,26 @@ class HiddenColumns extends BasePlugin {
    * @param {HTMLElement} TH Header's TH element.
    */
   onAfterGetColHeader(column, TH) {
-    if (this.isHidden(column)) {
+    if (!this.settings.indicators || column === -1) {
       return;
     }
 
-    let firstSectionHidden = true;
-    let i = column - 1;
-
-    do {
-      if (!this.isHidden(i)) {
-        firstSectionHidden = false;
-        break;
-      }
-      i -= 1;
-    } while (i >= 0);
+    const physicalColumn = this.hot.toPhysicalColumn(column);
+    const sequenceOfPhysicalIndexes = this.hot.columnIndexMapper.getIndexesSequence();
+    const currentPositionInSourceIndexes = sequenceOfPhysicalIndexes.indexOf(physicalColumn);
+    const classList = [];
 
-    if (firstSectionHidden) {
-      addClass(TH, 'firstVisibleColumn');
+    if (column >= 1 &&
+      this.isHidden(this.hot.toVisualColumn(sequenceOfPhysicalIndexes[currentPositionInSourceIndexes - 1]))) {
+      classList.push('afterHiddenColumn');
     }
 
-    if (!this.settings.indicators) {
-      return;
-    }
-
-    if (this.isHidden(column - 1)) {
-      addClass(TH, 'afterHiddenColumn');
-    }
-
-    if (this.isHidden(column + 1) && column > -1) {
-      addClass(TH, 'beforeHiddenColumn');
+    if (currentPositionInSourceIndexes < sequenceOfPhysicalIndexes.length - 1 &&
```

Comment:
sourceColumnsCount or something like that? We have `getNumberOfIndexes`
```suggestion
    if (currentPositionInSourceIndexes < sequenceOfPhysicalIndexes.length - 1 &&
```

Reply:
As above.